pkg/controller/nodelifecycle/node_lifecycle_controller.go

```diff
@@ -999,7 +1000,7 @@ func isNodeExcludedFromDisruptionChecks(node *v1.Node) bool {
 
 // tryUpdateNodeHealth checks a given node's conditions and tries to update it. Returns grace period to
 // which given node is entitled, state of current and last observed Ready Condition, and an error if it occurred.
-func (nc *Controller) tryUpdateNodeHealth(ctx context.Context, node *v1.Node) (time.Duration, v1.NodeCondition, *v1.NodeCondition, error) {
+func (nc *Controller) tryUpdateNodeHealth(ctx context.Context, node *v1.Node) (time.Duration, v1.NodeCondition, *v1.NodeCondition, bool, error) {
```

Comment:
I was wondering why the addition of bool is needed here. Thanks in advance for the clarification!

Reply:
In the original code, here were semantically misleading, giving the impression that markPodNotReady could be triggered if node switches status from True to non-True.  The lack of semantic clarity actually causes this bug.

https://github.com/kubernetes/kubernetes/blob/0f8fb9f916d499137f16424c6891cf48547617b2/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L878

In fact, markPodNotReady is only called here when the node switches from Ready to Unknown, and these details are hidden in tryUpdateNodeHealth. For example, you might be surprised that markPodNotReady is not called when the node switches from Ready to False (But I have no intention of fixing this problem, see the comments above).

We can trace the code and this conditional judgment is actually fully bound to tryUpdateNodeHealth, and only occurs if the Kubelet timeout to report a heartbeat.

I hope the code here is more semantic and readable, so that the next time you modify it, you'll understand where markPodNotReady was decided. So I return this variable in the function and you can know I want to markPodsNotReady.